src/number_theory/basic.lean

```diff
@@ -41,4 +50,247 @@ begin
+begin
+  cases n,
+  { rw [pow_zero, pow_zero, nat.cast_zero, mul_zero, add_zero] },
+  rw [nat.succ_sub_one, add_pow],
+  iterate 2 { rw [finset.range_succ, finset.sum_insert finset.not_mem_range_self] },
+  rw [nat.sub_self, pow_zero, nat.choose_self, mul_one, nat.cast_one, mul_one,
+    ← nat.add_one, nat.add_sub_cancel_left, pow_one,
+    nat.choose_succ_self_right, ← add_assoc, add_right_eq_self],
+  apply finset.sum_eq_zero,
+  intros i hi,
+  have h : (0 : R) ∣ p ^ (n + 1 - i) := hp2 ▸ pow_dvd_pow _
+    (nat.le_sub_left_of_add_le $ nat.succ_le_succ $ finset.mem_range.mp hi),
+  rw [mul_pow, eq_zero_of_zero_dvd h, zero_mul, mul_zero, zero_mul]
+end
+
+private lemma aux₂ (hp : p % 2 = 1) :
+  geom_sum₂ (a + p * b) a p = p * a ^ (p - 1) :=
+calc  ∑ i in finset.range p, (a + p * b) ^ i * a ^ (p - 1 - i)
+    = ∑ i in finset.range p,
+        (a ^ i * a ^ (p - 1 - i) + b * p * (a ^ ((i - 1) + (p - 1 - i)) * i)) :
+  finset.sum_congr rfl $ λ i hi, by rw [add_mul_pow_eq hp2, pow_add]; ring
+... = p * a ^ (p - 1) + b * a ^ (p - 2) * ↑((p - 1) / 2) * ↑p ^ 2 :
+  by rw [finset.sum_add_distrib, ← geom_sum₂, geom_sum₂_self,
+    ← finset.mul_sum, aux₁, ← finset.mul_sum, ← nat.cast_sum, finset.sum_range_id,
+    nat.mul_div_assoc _ (nat.dvd_sub_of_mod_eq hp), nat.cast_mul]; ring
+... = p * a ^ (p - 1) :
+  by rw [hp2, mul_zero, add_zero]
+
+end helpers
+
+end comm_ring
+
+namespace multiplicity
+
+section integral_domain
+
+open ideal ideal.quotient
+
+variables {R : Type*} [integral_domain R] [@decidable_rel R (∣)]
+
+lemma pow_sub_pow' {p : R} (hp : prime p)
+  {x y : R} (hxy₁ : p ∣ x - y) (hxy₂ : ¬p ∣ x * y)
+  {n : ℕ} (hn : ¬p ∣ n) :
+  multiplicity p (x ^ n - y ^ n) = multiplicity p (x - y) :=
+by rw [← geom_sum₂_mul, multiplicity.mul hp,
+  geom_sum₂_eq_zero hp hxy₁ hxy₂ hn, zero_add]
+
+variables {p : ℕ} (hp : prime (p : R)) (hp1 : p % 2 = 1)
+  {x y : R} (hxy₁ : ↑p ∣ x - y) (hxy₂ : ¬↑p ∣ x * y)
+include hp hp1 hxy₁ hxy₂
+
+lemma geom_sum₂_eq_one : multiplicity ↑p (geom_sum₂ x y p) = 1 :=
+begin
+  apply multiplicity.eq_some_iff.mpr,
+  split,
+  { rw ← mem_span_singleton at *,
+    rw ← ideal.quotient.eq at hxy₁,
+    rw [pow_one, ← eq_zero_iff_mem, ring_hom.map_geom_sum₂, hxy₁, geom_sum₂_self],
+    apply mul_eq_zero_of_left,
+    rw [← ring_hom.map_nat_cast (mk _), eq_zero_iff_mem, mem_span_singleton] },
+  { let I : ideal R := span {p ^ 2},
+    have hp' : (p : I.quotient) ^ 2 = 0,
+    { rw [← ring_hom.map_nat_cast (mk I), ← ring_hom.map_pow,
+        eq_zero_iff_mem, mem_span_singleton] },
+    cases hxy₁ with k hk,
+    rw [← mem_span_singleton, ← eq_zero_iff_mem, ring_hom.map_geom_sum₂,
+      sub_eq_iff_eq_add'.mp hk, ring_hom.map_add, ring_hom.map_mul, ring_hom.map_nat_cast,
+      aux₂ hp' hp1, ← ring_hom.map_nat_cast (mk I), ← ring_hom.map_pow, ← ring_hom.map_mul,
+      eq_zero_iff_mem, mem_span_singleton, pow_two, mul_dvd_mul_iff_left hp.1],
+    intro h,
+    exact hxy₂ (dvd_mul_of_dvd_right (hp.dvd_of_dvd_pow h) _) }
```

Comment:
I have the feeling that this could be easier, without quite as many `map_nat_cast`s at least, if we could work directly with the coercion `ℕ → I.quotient`. But adding that is definitely out of scope for this PR.